### PR TITLE
Integrate FlashAttention into HF OPT 

### PR DESCRIPTION
**Pull request by erichan1**
Integrate FlashAttention. 
 - Requires https://github.com/pytorch/pytorch/pull/81434 to work. torch._scaled_dot_product_attention is only there. 
 - Turn on fast path or go back to slow path using fast_attention=True/False flag. 
 - Turn on causal mask or turn it off for the fast attention path using fast_attention_causal = True/False. 
 - Does not support attention mask or padding mask on the fast path. 
 - Currently requires us to do an unnecessary conversion to Nestedtensor and back because the current FlashAttn implementation only takes NestedTensor. Will remove once torch._scaled_dot_product_attention supports regular tensor.

### CODE REVIEW

File: src/transformers/models/opt/modeling_opt.py [hunk 1]
```diff
@@ -191,71 +193,88 @@ def forward(
...
-        query_states = self._shape(query_states, tgt_len, bsz).view(*proj_shape)
-        key_states = key_states.view(*proj_shape)
-        value_states = value_states.view(*proj_shape)
+        if fast_attention:
+            key_states_fast = key_states.view(bsz, -1, self.embed_dim)
+            value_states_fast = value_states.view(bsz, -1, self.embed_dim)
+            seqlen = key_states_fast.shape[1]
+            query_states_fast = torch.nested_tensor(torch.unbind(query_states, dim=0))
+            key_states_fast = torch.nested_tensor(torch.unbind(key_states_fast, dim=0))
+            value_states_fast = torch.nested_tensor(torch.unbind(value_states_fast, dim=0))
```
Comment (jbschlosser) [inline src/transformers/models/opt/modeling_opt.py:202]: won't this result in padding within the nested tensors?

i.e. if query_states is a padded rectangular tensor, calling unbind on it will produce sequences padded to the same length, so we won't be taking advantage of nested tensors to reduce padding. Or am I missing something?

Comment (erichan1) [reply src/transformers/models/opt/modeling_opt.py:202]: I'm assuming only 0 padding tensor inputs for now. This is a hack just to make those tensors into NestedTensors because currently FlashAttention SDP requires NestedTensor. If FlashAttn SDP supported regular tensor I would just remove these entirely.

Comment (jbschlosser) [reply src/transformers/models/opt/modeling_opt.py:202]: gotcha, thanks for the clarification :)

**Comment by fzyzcjy** [comment]
Hi, is there any updates? Coming from https://github.com/HazyResearch/flash-attention/blob/main/usage.md

**Comment by puyuanOT** [comment]
Looking forward to the update!

**Comment by erichan1** [comment]
> Looking forward to the update!

Hey there @puyuanOT! Not working on this actively anymore. Check out [torch SDP](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#:~:text=Scaled%20dot%20product%20attention%20attempts,for%20enabling%20and%20disabling%20implementations.) to use FlashAttn in native torch!

**Comment by puyuanOT** [comment]
Thanks @erichan1 ! I will check it out.

**Comment by vincentmin** [comment]
@erichan1 Could you explain the reason for stopping to work on this feature? I think it would be a great implementation for the transformers library.
Regarding the torch SDP link, could you give instructions on how to use this torch feature when using a model in Huggingface transformers?

**Comment by vincentmin** [comment]
@erichan1 @amyeroberts Thank you for the clarifications. I now understand that BetterTransformer should offer the features I am looking for. I encourage you to write a blog post on Huggingface to advertise this to the world!

**Comment by younesbelkada** [comment]
Hi @erichan1 @amyeroberts @vincentmin 
This is correct, SDPA is now part of the optimum's `BetterTransformer` API, however this is only available for decoder-based models right now. 
We are indeed panning to write a blogpost soon with Pytorch to publicly announce the feature soon. We will keep you posted here!

**Comment by KatarinaYuan** [comment]
Hi, any recent updates on this blogpost for `BetterTransformer` that you mentioned earlier?

**Comment by younesbelkada** [comment]
Hi @KatarinaYuan 
Yes the blogpost is out and is here: https://pytorch.org/blog/out-of-the-box-acceleration/

**Comment by KatarinaYuan** [comment]
Thank you!

**Comment by ASR-SCI** [comment]
I use the transformer trainer + FSDP llama training options, model cannot be saved, and unable to use bettertransformer.reverse() convert to  original model. I don't know how to deal with this problem.

**Comment by EwoutH** [comment]
Are there any updates on the integration of FlashAttention into HuggingFace Transformers?

**Comment by younesbelkada** [comment]
@EwoutH 
Flashattention should be used as a backend for torch.SDPA which is itself integrated into `BetterTransformer` API. Make sure to install the latest transformers and optimum libraries and run:
```python
model = model.to_bettertransformer()
```

**Comment by tmm1** [comment]
is BetterTransformer up to date with FlashAttention v2?

**Comment by fxmarty** [comment]
Hi, BetterTransformer integrates with PyTorch SDPA (for now), and PyTorch has not integrated flash v2 yet: https://github.com/pytorch/pytorch/pull/105602. Hopefully it will be there in Pytorch 2.1.
